### PR TITLE
Update dependencies, refactor styles and scripts, add some goodies

### PR DESCRIPTION
**Pull request by rdeago**
Since #15 and #16 are unfortunately out of date, here's an update to current versions.

**EDIT:** I kinda got carried away 😁 so here are more changes:

- Update styles and scripts from [upstream](https://github.com/StartBootstrap/startbootstrap-clean-blog) where they have undergone some serious refactoring.
- Refactor styles to better separate upstream styles from modifications and expand override possibilities (e.g. make customization of Bootstrap variables possible by just adding one input file in a site).
- Remove styles, scripts and dependencies that were only used by the contacts page.
- Refactor what little JavaScript remains to not use jQuery.
- Remove jQuery as neither Bootstrap 5 nor upstream code depends on it any more.
- Add better styling for blockquotes (see #22) blockquote
- Add pluggable comment engines and Giscus support (see #24)
- Add an `IsPage` boolean setting for pages ("About" and the like). Pages should be put in a `pages` folder and are added to the navigation bar by default.
- Set `Order` to 98 for Posts and 99 for Tags, thus establishing a clear rule for navbar links: pages with an `Order` of at least 100 will be linked after archives.
- Rename `_post-footer` partial to `_post-after-content` as it is not actually included in the footer.
- Add a `_page-after-content` partial for consistency with `_post-after-content`.
- Add a `_common-after-content` partial for _all_ pages, included just after `_post-after-content` in posts and just after `_page-after-content` in pages.
- Move the navbar's `<ul>` tag in `_navigation.cshtml`, leaving only the items in `_navbar.cshtml`
- Make the navbar's search field a proper navbar item, so that there's a correct margin between it and the last navbar link.
- Add a `.editorconfig` file to the repository for better collaboration.
- Remove unused images (i.e. the whole `img` folder).
- Remove redundant favicon link.
- Remove unused sidebar styles.
- Make the masthead gradient configurable.
- Add support for an `ImageAttribution` setting. If it is present in any page, a nice semi-transparent badge is created in the lower-right corner of the masthead, with the value of `ImageAttribution` inserted as raw HTML. Makes displaying proper attribution for images very easy, for example when using photos from [Unsplash](https://unsplash.com).
- Make the copyright notice fully customizable, so for example proper CC licensing can be applied without having to override the whole footer.
- Add support for a `SiteDescription` global setting to set the home page subtitle.

### CODE REVIEW

**Comment by rdeago** [comment]
@daveaglick if I promise to stop here 😁 will you take a look at this PR? 🙏

I know you're very busy, but hey, nobody can say I didn't try.

If there are some changes you don't approve, just say so and I'll revert them in a snap.

Thanks again for your amazing work!

**Comment by daveaglick** [comment]
Whoah! This is some amazing work. I'll make every effort to take a closer look soon and get it merged. Thankfully I'm less worried about breakage and backwards compatibility in the themes than in the actual libraries - I figure themes are going to evolve naturally and folks can pick them up at whatever commit makes sense for them (or not backport at all if they're already up and running). So mainly I'll just be looking to make sure everything still builds and works correctly with the various edge cases I've seen over the years.

Stay tuned!

**Comment by rdeago** [comment]
> This is some amazing work.

Thanks! 😊 A lot of merit goes to the StartBootstrap team, anyway, mainly for their work on SASS files.

> Thankfully I'm less worried about breakage and backwards compatibility in the themes than in the actual libraries

I figured so, otherwise I'd have taken a lot less liberties. 😉

By the way, on second thought I'm not sure that adding the `IsArchive` setting was the right thing to do. More on this in an incoming issue.

**Comment by rdeago** [comment]
Whew! OK, _now_ I'm going to stop. Promise. 😅 I'll just update README.md to document the changes.

There's still one thing that bugs me: I had to special-case the Posts page by adding an `IsPostArchive` setting to it and checking for it in Razor every time I have to tell whether a document is a post. If I just set `IsPost` to `false` in the front matter of `posts/index.cshtml`, the page disappears completely from deployment. I guess it has to do with some interaction between the `Content` and `Archives` pipelines, but I couldn't find a cleaner way to solve it, so here we go.

**Comment by rdeago** [comment]
There's still room for improvement, for example:

- add optional support for [Mermaid](https://mermaid-js.github.io/mermaid/#/) (but I'm not sure whether it requires some trickery at the Statiq.Web level);
- make styles more customizable (but it requires that upstream set their SASS variables in a friendlier way, using `!default` and such);
- add support for multiple authors, including an Authors archive and a convention for author avatars (e.g. `img/authors/author-name.png`, or configure author data in a JavaScript file and use Gravatar) but frankly, who would even use it?
- get rid of both upstream and Bootstrap completely, use plain semantic HTML5 + SASS, adding switchable and user-configurable layouts _à la_ [CSS Zen Garden](http://www.csszengarden.com), thus making this the only Statiq blog theme one could possibly ever need... but that's another project entirely 😁 and I wish I had the time to pursue it but I don't. Maybe when I retire.

**Comment by rdeago** [comment]
> Whew! OK, _now_ I'm going to stop. Promise.

I know, I know... 🙄 

_(goes back to work on the README)_

**Comment by rdeago** [comment]
Aaaaand, done! I mean, the README could use some more love, but I had to stop _somewhere_.

@daveaglick now it's review time for real.

**Comment by rdeago** [comment]
Live example of a blog using this PR: https://riccar.do

Source code: https://github.com/rdeago/rdeago.github.io

**Comment by atiq-cs** [comment]
@daveaglick how can we help you to review this PR since it's a large one? Please let us know.

**Comment by daveaglick** [comment]
@atiq-cs Thanks for asking! I'll take the recommendation of another community member since this a theme and not part of the core (I.e. I don't feel too bad if something in a theme breaks since it's not quite the same rigor as a library 😂). If you have the time and wouldn't mind taking a look and seeing if it all works as expected, I'd feel more comfortable with a YOLO merge.

**Comment by rdeago** [comment]
> I don't feel too bad if something in a theme breaks since it's not quite the same rigor as a library 😂

While some breaking changes were inevitable (there was still some stuff dated 2014 or so - early Renaissance period, in other words 😂) I must confess it's been both fun and stress-relieving to be allowed to break things, elephant-in-china-shop style. 😁

@atiq-cs please feel free to open as many conversations as you want, even on points you're just curious about. Resolved conversations may make for some of the best possible documentation.

**Review by atiq-cs — APPROVED** [review]

**Comment by atiq-cs** [comment]
Sorry, I meant to approve the first commit and comment that it looks good. Unfortunately, it has approved all. In next comments, I will just provide review on each commit.

**Comment by atiq-cs** [comment]
Sorry, this is a bit more work for you. However, do you mind separating out the commits in separate PRs? @rdeago 

It makes the review easier and changes comprehensible.

**Review by atiq-cs — CHANGES_REQUESTED** [review]
Meant to approve following commits, however, it approved all the commits instead,
- 53ad865
- 0dc00e5

Sorry, this is a bit more work for you. However, do you mind separating out the commits in separate PRs? @rdeago

It makes the review easier and changes comprehensible.

For example, do you mind starting with "Bump up jQuery"?

**Comment by rdeago** [comment]
@atiq-cs while I'll admit this PR is not as clean as I'd like it to be, there is hardly any "work-in-progress-do-not-use" commit. GitHub makes reviewing a single commit just as easy as reviewing them all together, and you can comment on each single line of code, or group thereof.

Of course I can separate out changes. I'd do some things differently of course. For example: bump dependencies one by one, then all aesthetic code fixes in one swallow, then add new features one by one; changes due to the differences between Bootstrap 4 and 5 should go together with the Bootstrap version bump; skip https://github.com/statiqdev/CleanBlog/pull/23/commits/7edfaf524aaa1e171919198e179ac1c8608c7ee8 entirely so https://github.com/statiqdev/CleanBlog/pull/23/commits/875b1f190f21b274b017560761572d8b047b877a doesn't have to undo it (that's the fix to #25); and so on. This would leave us with 20 to 25 commits (maybe 30 because https://github.com/statiqdev/CleanBlog/pull/23/commits/d0b861f74daef53685743e23f1c172f3858036ff can be further split) instead of 38 and a much cleaner history.

It's just going to take time, as I'm currently busy as a bee.

**Comment by daveaglick** [comment]
I don't want to drag this out any longer than we need to since it's already been on the shelf for a while. @atiq-cs I really appreciate your willingness to do a full review! Is there a middle ground though where you'd be able to just run through a generation with the new theme and see if anything sticks out without requiring the extra effort to split it all back up again? I'll endeavor to do the same tomorrow.

I don't know that we have to verify every change, just make sure it's not going to blow up on new users. If we've passed that minimum bar I'm okay merging it in - we can always iterate and/or fix things as they come later.

Thanks both!

**Comment by rdeago** [comment]
Should it be of any help, here's my blog's source:
https://github.com/rdeago/rdeago.github.io/tree/main/web
It uses practically all of the features in this PR.

**Comment by rdeago** [comment]
Quick guide to updating a pre-this-PR blog:

- either move pages to a `pages` folder, or add `IsPage: true` to the front matter of each;
- if you have a `_post-footer.cshtml` file, rename it to `_post-after-content.cshtml` - if you did it just to add Giscus, consider using the Giscus comment engine instead (info in [my fork's README](https://github.com/rdeago/Statiq-CleanBlog/tree/update-dependencies#post-comments));
- if you have a `_navbar.cshtml`, delete `<ul>` and `</ul>` from it, leaving only the `<li>` tags;
- if you overrode `_footer.cshtml` just to add a copyright notice, consider using the `Copyright` global setting instead (if it's just a string) and/or overriding `_copyright.cshtml`, [like this](https://github.com/rdeago/rdeago.github.io/blob/main/web/input/_copyright.cshtml);
- overridden SCSS is a bit more involved, but in most cases you shouldn't need to tweak existing code: `scss/_overrides.scss` is the right place for adding styles.

**Comment by atiq-cs** [comment]
> you'd be able to just run through a generation with the new theme

Sure. I will test with a fresh check out. What breakages do you care about that I should look out for? @daveaglick 

I plan to check following,
- test the main page and individual post
- test archive page
- test tags
- test search feature

**Comment by atiq-cs** [comment]
I pulled in this PR and ran locally. Looks about the same. I only tried things I mentioned in previous comment.

- [Home Page Output Image](https://user-images.githubusercontent.com/7858031/205252505-7624eddc-0fb6-45bf-a239-57bf81f01741.png)
- [Individual Post page  Output Image](https://user-images.githubusercontent.com/7858031/205252531-076d4e94-b97a-4767-ad9c-0f0d7cbf5507.png)
- [Posts Output Image](https://user-images.githubusercontent.com/7858031/205252557-87bf8142-8639-44bc-94b3-43db32167267.png)
- [Tags Output Image](https://user-images.githubusercontent.com/7858031/205252585-ec0bc313-46fd-4fbf-8611-b64a9d91ab71.png)

ref, https://github.com/atiq-cs/statiq-template

I haven't tried sophisticated features like giscus for example. @daveaglick please let me know what else you are interested in, would love to test more.

**Comment by rdeago** [comment]
@atiq-cs your template uses a `Description` global setting for the home page subtitle. That should be `SiteDescription` now. The rationale is to not force a `Description` value, that may only make sense when displayed on the home page, to all pages lacking a `Description` setting of their own.

A couple more things you may try if you feel like it:

- Blockquotes in post should be styled as shown in original post above.
- An "About" page (`pages/about.md`) should show in top navigation bar before "Posts" - unless it has an `Order` setting >= 100, in which case it should appear after "Tags".
- A post outside of the `posts` folder should not be included in archives, unless it has an `IsPost` setting of `true`.
- A page outside of the `pages` folder should not be included in the top navbar, unless it has an `IsPage` setting of `true`.
- Adding a `scss/_variable-overrides.scss` file with the following content should make the masthead gradient red and more contrasted:
  ```yaml
  $masthead-main-color: $red;
  $masthead-gradient-spread: 60%;
  ```
- Use a masthead image with attribution. For example, add this to the front matter of any page or post (no `/img` folder required thanks to Pixabay's CDN):
  ```yaml
  Image: https://cdn.pixabay.com/photo/2022/11/25/16/30/brown-hairstreak-7616422_960_720.jpg
  ImageAttribution: Photo by <a href="https://pixabay.com/users/erik_karits-15012370">Erik_Karits</a> on <a href="https://pixabay.com/photos/brown-hairstreak-butterfly-insect-7616422">Pixabay</a>
  ```

**Review by atiq-cs — APPROVED** [review]
Thanks for addition of giscus comment engine support.
- 1ef8bf0

**Comment by atiq-cs** [comment]
I am okay with YOLO merge @daveaglick Please let me know if you want me to check anything else.

**Comment by rdeago** [comment]
The best test IMHO would be someone with no previous knowledge of Statiq or even Razor to go from zero to blog without having to jump through too many hoops. (Don't get me wrong though, I'm glad and grateful for the things I had to learn.)

What hoops, you may ask... it's not rocket science after all. Well, think image attribution for example: where do I put it? At the top of a post? At the bottom? In the alt text? Oh wait, can't do that last one... Or, how do I add a CC license? Overriding the whole footer seems a bit overkill (not to mention prone to update hell); `_post-footer.cshtml` is nice, but what about pages? And so on.

In short, I tried to solve some small annoyances that, to be honest, are common to all blog templates I've ever tried. If anyone ever chooses Statiq and CleanBlog mostly because of the "out-of-the-box experience", I will consider it a victory. If not, hey, at least my own blog came out nice, I learned something new, and I got to share back - so there's that, and to me it's far from nothing. Thanks @daveaglick and @atiq-cs for making this possible.

**Comment by daveaglick** [comment]
No, thank you @rdeago! I've also finally successfully played around with this on my own blog and it works great. Between that and @atiq-cs running it, I think we're good. I _really_ appreciate all the work you put into this. If it's not yet obvious, my time has been at a premium and so I've been unable to really keep this theme updated like it should be (not to mention best laid plans for some new themes - someday hopefully). It's really nice to know the community is willing to support the project.

**Comment by atiq-cs** [comment]
**jquery**
Comment regarding commit# 43ba7cb00,

Interesting: I see you removed `input/vendor/jquery/jquery.js` but you updated 'jquery-min'. So, that files is not used at all? The blog only consumes 'jquery min'? Thanks for the optimization if that's the case. @rdeago 

**bootstrap**
Seeing similar change in ca33d33b7, only to keep min version.

**Font Awesome**
I am seeing similar change (`all.css` and a bunch other probably unused css are removed) with font awesome in commit# 0dc00e5c9.

Guessing these all above are for optimization and clean up purposes.

(just checking out of curiosity)
